File: widgetssdk/src/main/java/com/glia/widgets/chat/domain/IsShowSendButtonUseCase.java
package com.glia.widgets.chat.domain;

import com.glia.widgets.core.engagement.GliaEngagementRepository;
import com.glia.widgets.fileupload.FileAttachmentRepository;

public class IsShowSendButtonUseCase {
    private final GliaEngagementRepository engagementRepository;
    private final FileAttachmentRepository fileAttachmentRepository;

    public IsShowSendButtonUseCase(
            GliaEngagementRepository engagementRepository,
            FileAttachmentRepository fileAttachmentRepository
    ) {
        this.engagementRepository = engagementRepository;
        this.fileAttachmentRepository = fileAttachmentRepository;
    }

    public boolean execute(String message) {
        return hasText(message) || hasEngagementOngoingAndReadyToSendUnsentAttachments();
    }

    private boolean hasText(String message) {
        return message != null && !message.isEmpty();
    }

    private boolean hasEngagementOngoingAndReadyToSendUnsentAttachments() {
        return engagementRepository.hasOngoingEngagement() &&
                fileAttachmentRepository.getReadyToSendFileAttachments().size() > 0;
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/chat/domain/IsShowSendButtonUseCase.kt
package com.glia.widgets.chat.domain

import com.glia.widgets.core.fileupload.EngagementFileAttachmentRepository
import com.glia.widgets.core.secureconversations.domain.ManageSecureMessagingStatusUseCase
import com.glia.widgets.engagement.domain.IsQueueingOrLiveEngagementUseCase

internal class IsShowSendButtonUseCase(
    private val isQueueingOrLiveEngagementUseCase: IsQueueingOrLiveEngagementUseCase,
    private val fileAttachmentRepository: EngagementFileAttachmentRepository,
    private val manageSecureMessagingStatusUseCase: ManageSecureMessagingStatusUseCase
) {
    operator fun invoke(message: String?): Boolean {
        return when {
            manageSecureMessagingStatusUseCase.shouldBehaveAsSecureMessaging -> hasText(message) || hadReadyToSendUnsentAttachments()
            else -> hasText(message) || hasEngagementOngoingAndReadyToSendUnsentAttachments()
        }
    }

    private fun hasText(message: String?): Boolean {
        return !message.isNullOrEmpty()
    }

    private fun hadReadyToSendUnsentAttachments(): Boolean {
        return fileAttachmentRepository.getReadyToSendFileAttachments().isNotEmpty()
    }

    private fun hasEngagementOngoingAndReadyToSendUnsentAttachments(): Boolean {
        return isQueueingOrLiveEngagementUseCase.hasOngoingLiveEngagement && hadReadyToSendUnsentAttachments()
    }
}
